sweep/sweeper.go

```diff
@@ -1,38 +1,690 @@
+		return nil
+
+	}
+
+	log.Tracef("Sweeper starting")
+
+	// Retrieve last published tx from database.
+	lastTx, err := s.cfg.Store.GetLastPublishedTx()
+	if err != nil {
+		return err
+	}
+
+	// Republish in case the previous call crashed lnd. We don't care about
+	// the return value, because inputs will be reoffered and retried
+	// anyway. The only reason we republish here is to prevent prevent the
+	// corner case where lnd goes into a restart loop because of a crashing
+	// publish tx where we keep deriving new output script. By publishing
+	// and possibly crashing already now, we haven't derived a new output
+	// script yet.
+	if lastTx != nil {
+		log.Debugf("Publishing last tx %v", lastTx.TxHash())
+		s.cfg.PublishTransaction(lastTx)
+	}
+
+	// Retrieve relay fee for dust limit calculation. Assume that this will
+	// not change from here on.
+	s.relayFeePerKW = s.cfg.Estimator.RelayFeePerKW()
+
+	// Register for block epochs to retry sweeping every block.
+	bestHash, bestHeight, err := s.cfg.ChainIO.GetBestBlock()
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("Best height: %v", bestHeight)
+
+	blockEpochs, err := s.cfg.Notifier.RegisterBlockEpochNtfn(
+		&chainntnfs.BlockEpoch{
+			Height: bestHeight,
+			Hash:   bestHash,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	// Start sweeper main loop.
+	s.wg.Add(1)
+	go func() {
+		err := s.collector(blockEpochs.Epochs, bestHeight)
+		if err != nil {
+			log.Errorf("Sweeper stopped: %v", err)
+		}
+
+		defer blockEpochs.Cancel()
+		defer s.wg.Done()
+	}()
+
+	return nil
+}
+
+// Stop stops sweeper from listening to block epochs and constructing sweep
+// txes.
+func (s *UtxoSweeper) Stop() error {
+	if !atomic.CompareAndSwapUint32(&s.stopped, 0, 1) {
+		return nil
+	}
+
+	log.Tracef("Sweeper shutting down")
+
+	close(s.quit)
+	s.wg.Wait()
+
+	log.Tracef("Sweeper shut down")
+
+	return nil
+}
+
+// SweepInput sweeps inputs back into the wallet. The inputs will be batched and
+// swept after the batch time window ends.
+//
+// NOTE: Extreme care needs to be taken that input isn't changed externally.
+// Because it is an interface and we don't know what is exactly behind it, we
+// cannot make a local copy in sweeper.
+func (s *UtxoSweeper) SweepInput(input Input) (chan Result, error) {
+	if input == nil || input.OutPoint() == nil || input.SignDesc() == nil {
+		return nil, errors.New("nil input received")
+	}
+
+	log.Debugf("Sweep request received: %v", input.OutPoint())
+
+	sweeperInput := &sweepInputMessage{
+		input:      input,
+		resultChan: make(chan Result, 1),
+	}
+
+	// Deliver input to main event loop.
+	select {
+	case s.newInputs <- sweeperInput:
+	case <-s.quit:
+		return nil, fmt.Errorf("sweeper shutting down")
+	}
+
+	return sweeperInput.resultChan, nil
+}
+
+// collector is the sweeper main loop. It processes new inputs, spend
+// notifications and counts down to publication of the sweep tx.
+func (s *UtxoSweeper) collector(blockEpochs <-chan *chainntnfs.BlockEpoch,
```

Comment:
I am in favor of strict errors. Take for example the wallet problems we had. Those were producing errors, but went unnoticed because they were logged and stepped over.